src/components/reports/cost.ts

```diff
@@ -0,0 +1,169 @@
+
+import costReportTemplate from './cost.njk';
+
+interface ICostable {
+  readonly incVAT: number;
+  readonly exVAT: number;
+}
+
+type ICostSummary = ICostable;
+
+interface IOrgCostRecord extends ICostable {
+  readonly orgGUID: string;
+  readonly orgName: string;
+
+  readonly quotaGUID: string;
+  readonly quotaName: string;
+}
+
+interface IQuotaCostRecord extends ICostable {
+  readonly quotaGUID: string;
+  readonly quotaName: string;
+}
+
+export async function viewCostReport(
+  ctx: IContext,
+  params: IParameters,
+): Promise<IResponse> {
+  const rangeStart = moment(params.rangeStart, 'YYYY-MM-DD').toDate();
+  const rangeStop  = moment(rangeStart).add(1, 'month').toDate();
+
+  const billingClient = new BillingClient({
+    apiEndpoint: ctx.app.billingAPI,
+    accessToken: ctx.token.accessToken,
+  });
+
+  const cf = new CloudFoundryClient({
+    accessToken: ctx.token.accessToken,
+    apiEndpoint: ctx.app.cloudFoundryAPI,
+  });
+
+  const orgs = await cf.organizations();
+  const orgGUIDs = orgs.map(o => o.metadata.guid);
+
+  const orgQuotas: {[key: string]: IOrganizationQuota} = (await Promise.all(
+    orgs.map(async (org: IOrganization) => {
+      const quota = await cf.organizationQuota(org.entity.quota_definition_guid);
+      return {[org.metadata.guid]: quota};
+    }),
+  ))
+  .reduce(/* istanbul ignore next */ (accumulator, quota) => {
+    return { ...accumulator, ...quota };
+  });
+
+  const billableEvents = await billingClient.getBillableEvents({
+    rangeStart, rangeStop, orgGUIDs,
+  });
+
+  const orgBillableEvents = aggregateBillingEvents(billableEvents);
+  const orgCostRecords = createOrgCostRecords(
+    orgs,
+    orgQuotas,
+    orgBillableEvents,
+  );
+  const quotaCostRecords = createQuotaCostRecords(orgCostRecords);
+
+  const totalBillables = sumRecords(billableEvents);
+  const billableEventCount = billableEvents.length;
+
+  return {
+    body: costReportTemplate.render({
+      date: moment(rangeStart).format('MMMM YYYY'),
+      orgCostRecords: Object.values(orgCostRecords),
+      quotaCostRecords,
+      totalBillables,
+      billableEventCount,
+    }),
+  };
+}
+
+export function aggregateBillingEvents(
+  billableEvents: ReadonlyArray<IBillableEvent>,
+): {[key: string]: ReadonlyArray<IBillableEvent>} {
+  return billableEvents
+    .reduce(
+      (
+        accumulator: {[key: string]: ReadonlyArray<IBillableEvent>},
+        billableEvent: IBillableEvent,
+      ) => {
+        const accEvents = accumulator[billableEvent.orgGUID] || [];
+        return {
+          ...accumulator,
+          [billableEvent.orgGUID]: accEvents.concat([billableEvent]),
+        };
+      }, {});
+}
+
+export function createOrgCostRecords(
+  orgs: ReadonlyArray<IOrganization>,
+  orgQuotas: {[key: string]: IOrganizationQuota},
+  orgBillableEvents: {[key: string]: ReadonlyArray<IBillableEvent>},
+): ReadonlyArray<IOrgCostRecord> {
+  return orgs
+  .map(org => {
```

Comment:
I wonder if it is worth it to iterate across all events and aggregate all the cost for non existent orgs and quotas